File: scripts/usf-proxy.py
#!/usr/bin/env python3
"""
USF rippled proxy
Run on ripple.cs.usfca.edu: python3 usf-proxy.py
Forwards HTTP POST requests to the local rippled node at 127.0.0.1:5005
"""

import json
import urllib.request
from http.server import HTTPServer, BaseHTTPRequestHandler

RIPPLED_URL = 'http://127.0.0.1:5005'
PORT = 3001

class ProxyHandler(BaseHTTPRequestHandler):

    def log_message(self, format, *args):
        print(f"{self.address_string()} - {format % args}")

    def do_OPTIONS(self):
        self.send_response(200)
        self._set_cors()
        self.end_headers()

    def do_POST(self):
        length  = int(self.headers.get('Content-Length', 0))
        body    = self.rfile.read(length)

        try:
            req = urllib.request.Request(
                RIPPLED_URL,
                data=body,
                headers={'Content-Type': 'application/json'},
                method='POST'
            )
            with urllib.request.urlopen(req, timeout=10) as resp:
                data = resp.read()
            self.send_response(200)
            self._set_cors()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(data)

        except Exception as e:
            print(f"Error: {e}")
            self.send_response(502)
            self._set_cors()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': str(e)}).encode())

    def _set_cors(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')

if __name__ == '__main__':
    server = HTTPServer(('0.0.0.0', PORT), ProxyHandler)
    print(f"USF rippled proxy running on port {PORT}")
    print(f"Forwarding to {RIPPLED_URL}")
    server.serve_forever()

File: projects/eviair/scripts/usf-proxy.js
// scripts/usf-proxy.js
// Run this on ripple.cs.usfca.edu: node usf-proxy.js
// It proxies HTTP POST requests to the local rippled node at 127.0.0.1:5005

const http = require('http')

const PORT = 3001

const server = http.createServer(async (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*')
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS')
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type')

  if (req.method === 'OPTIONS') {
    res.writeHead(200)
    res.end()
    return
  }

  if (req.method !== 'POST') {
    res.writeHead(405)
    res.end('Method not allowed')
    return
  }

  let body = ''
  req.on('data', chunk => body += chunk)
  req.on('end', () => {
    const options = {
      hostname: '127.0.0.1',
      port:     5005,
      path:     '/',
      method:   'POST',
      headers:  {
        'Content-Type':   'application/json',
        'Content-Length': Buffer.byteLength(body)
      }
    }

    const proxy = http.request(options, (rippleRes) => {
      let data = ''
      rippleRes.on('data', chunk => data += chunk)
      rippleRes.on('end', () => {
        res.writeHead(200, { 'Content-Type': 'application/json' })
        res.end(data)
      })
    })

    proxy.on('error', (e) => {
      console.error('Rippled error:', e.message)
      res.writeHead(502)
      res.end(JSON.stringify({ error: 'Could not reach rippled node' }))
    })

    proxy.write(body)
    proxy.end()
  })
})

server.listen(PORT, '0.0.0.0', () => {
  console.log(`USF rippled proxy running on port ${PORT}`)
  console.log(`Forwarding to local rippled at 127.0.0.1:5005`)
})
